Start documenting the SDK architecture (#342)
bitwarden/contributing-docs: docs/architecture/sdk/data-models.md

# Data Models

The SDK is expected to be used in a variety of projects, and to that end it defines a stable public

> interface. Care must be taken to avoid exposing the inner workings on the SDK including it's data

Reviewer: ```suggestion
interface. Care must be taken to avoid exposing the inner workings on the SDK including its data
```

model.

## Public interface

The public interface of the SDK is anything that is exposed externally, be it function names, types,
or exposed models. The SDK generally exposes request and response models that contains the expected
data for each API call.

We can generally group the public models into the following categories:

> - **View models**: These are models that generally represents a decrypted state, examples are

Reviewer 2: ⛏️ There are a few "These are"s that could be omitted as that's implied.

  `CipherView`, `CipherListView`, `FolderView`, etc.
- **Request models**: These are models that are used to send data into the SDK. Some examples are
  `ProjectGetRequest`, `ProjectCreateRequest`, etc.

> - **Response models**: Returns data from the SDK, one example is `ProjectResponse`.

Reviewer 2: ```suggestion
- **Response models**: Returns data from the SDK e.g. `ProjectResponse`.
```

## Internal models

> The SDK also maintains internal models

Reviewer: ```suggestion
The SDK also maintains internal models:
```

For consistency with the `Public interface` section

> - **Api models**: These are [auto-generated models](./server-bindings.md) that are used to interact

Reviewer 2: ```suggestion
- **API models**: These are [auto-generated models](./server-bindings.md) that are used to interact
```
